core/internal/requestlist/request-list.go

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) 2018-2019 NEC Laboratories Europe GmbH.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package requestlist provides functionality to maintain a set of
+// Request messages.
+package requestlist
+
+import (
+	"sync"
+
+	"github.com/hyperledger-labs/minbft/messages"
+)
+
+// List defines methods to manipulate a set of Request messages. The
+// set is only capable of keeping a single message per client. All
```

Comment:
(Might be a stupid question) is "keeping a single message per client" enough? If we have multiple pending requests for a client, the older one seems to be just ignored. If you have a good reason to do it or have a plan to extend it later, that's ok.

Reply:
MinBFT paper and current implementation does not allow correct clients to submit multiple requests in parallel. The client has to wait for the last request to complete before submitting another request. Different clients, in contrast, can submit requests independently and asynchronously w.r.t. each other.

We can think of relaxing this requirement as a future optimization. In fact, we have this in [our list of features to consider](https://github.com/hyperledger-labs/minbft#roadmap): "Asynchronous requests: enabling parallel processing of requests".

Reply:
However, let me think about this detail a bit more. Now, I'm not sure if storing only the last pending request per client would be enough to use this component in view change...

Reply:
So any request that was executed will be present in a valid new-view certificate and will be propagated to the new view by the view change mechanism. That means, after view change is completed, consensus may only need to be restarted for the last client's request. Thus, it seems enough to keep a single message per client in the pending request list.

Reply:
Thanks, so it's clear to me now, and I agree with the proposed change.